enonic/src/main/resources/site/parts/comments-list/comments-list.ts

```diff
@@ -0,0 +1,230 @@
+  comments: ReadonlyArray<Content<Comment>>,
+  isAuthenticated: boolean,
+  path: string,
+  partConfig: CommentsListPartConfig,
+  parentPath: string,
+): ThymeleafParams {
+  return {
+    comments: createCommentsTree(userMapping, partConfig.headingTag)(comments),
+    serviceUrl: serviceUrl('comment'),
+    parentPath: parentPath,
+    pathSelf: path,
+    loginUrl: getLoginUrl(partConfig.loginPageId),
+    logOut: logoutUrl({}),
+    isAuthenticated
+  }
+}
+function getLoginUrl(partLoginPageId?: string) {
+
+  return app.config.loginPageId ?
+    pageUrl(app.config.loginPageId) :
+    partLoginPageId ?
+    pageUrl(partLoginPageId) :
+    loginUrl({});
+}
+
+function notNullOrUndefined(str: string | undefined): boolean {
+  return (str !== undefined) && (str !== null);
+}
+
+export const byParentFirst = pipe(
+  ordBoolean,
+  contramap((c: Content<Comment>) => notNullOrUndefined(c.data.parentId))
+)
+
+export function childrenToParent(acc: ReadonlyArray<ContentTree>, curr: ContentTree): ReadonlyArray<ContentTree> {
+  if (curr.parentId) {
+    return acc.map((comment: ContentTree) => {
+      if(comment._id === curr.parentId) {
+        const children = comment.children ?? []
+        return {...comment, children: [...children, curr]} as ContentTree
+      }
+      return comment
+    })
+  } else {
+    return [...acc, curr]
+  }
+}
+
+export interface ContentTree {
+  _id: string
+  title: string
+  body: string
+  createdTime: string
+  createdTimeDisplay: string
+  parentId?: string,
+  creatorDisplayName?: string;
+  children?: ReadonlyArray<ContentTree>
+}
+
+type HeadingTag = "h2" | "h3" | "h4" | "h5"
+const commentResponseMap: Record<HeadingTag, string> = {
+  h2: "h3",
+  h3: "h4",
+  h4: "h5",
+  h5: "h6"
+}
+
+export function createCommentsTree(userMapping: UserMapping, headingTag: HeadingTag): (c: ReadonlyArray<Content<Comment>>) => ReadonlyArray<ContentTree> {
+  return flow(A.sort(byParentFirst), A.map(createCommentTree(headingTag, userMapping)), A.reduce([], childrenToParent))
+}
+function createCommentTree (headingTag: HeadingTag, userMapping: UserMapping) {
+  return function mapCommentToContentTree(comment: Content<Comment>) {
+    const heading = createTemplateTag(headingTag, comment.data)
+    return {
+      _id: comment._id,
+      heading:  processHtml({value: heading}),
+      title: comment.data.title,
+      body: comment.data.body,
+      creatorDisplayName: userMapping[comment.creator],
+      parentId: comment.data.parentId,
+      createdTime: comment.createdTime,
+      createdTimeDisplay: pipe(
+        comment.createdTime,
+        parseISO,
+        formatWithOptions({}, 'dd.MM.yyyy')
+      )
+    }
+  }
+}
+
+function createTemplateTag (headingTag: HeadingTag, comment: Comment) {
+  const tag = mapCommentResponseHeading(headingTag, comment)
+  return`<${tag} class="${notNullOrUndefined(comment.parentId) ? 'comment-response-heading': 'comment-comment-heading'}">${comment.title}</${tag}>`
```

Comment:
This is a bit long, so I would maybe pull the class-building-ternary out, to make the line shorter.

I found it a bit difficult to see the difference between  `'comment-response-heading'` and `'comment-comment-heading'` so I would consider setting them to constants or an enum.

Btw, constants should be named with the least specific part first (to group constants together), and most specific parts last.

```typescript
const CLASS_HEADING_COMMENT = 'comment-comment-heading';
const CLASS_HEADING_RESPONSE = 'comment-response-heading';

// or as an enum

enum CssClass {
  HeadingComment = 'comment-comment-heading',
  HeadingResponse = 'comment-response-heading'
}

...
```